### PR TITLE
Simplify mage setup, fmt, test:ci

### PR DESCRIPTION
**Pull request by austinbyers**
## Background

CI has been reduced to a single main task: `mage test:ci`, which now runs in a minute on my laptop (90 seconds in CI)
We also remove several unneeded mage targets.

Closes: #457 

### Clean and Setup
All of the setup commands have been collapsed to a single (parallelized) `mage setup`. We had them previously separated for two reasons:

1. CI could run only the setup relevant for each individual task. We had originally planned to create many separate CI tasks, but after discussion decided instead to consolidate. Additionally, in the next PR, I will be leveraging [cached dependencies](https://circleci.com/docs/2.0/caching/) so that CI doesn't have to download all go/python/npm packages for every build
2. We isolated `setup:swagger` so users didn't have to install the other libraries if they weren't strictly needed for deployment. But they still have to spend a lot of time downloading all of the Go modules, the extra time for the other libraries is not significant when running in parallel. When deploying from source, it's reasonable that people would want to first run the tests, for example, to make sure everything works.

Other changes to setup:

* `go get` modules early on. Might as well do this in parallel with the rest of the setup instead of waiting until compilation
* Existing files are left alone if they are the correct version. Previously, `mage setup:all` would essentially remove most of `.setup` and re-download it. With the aforementioned CircleCI caching, the setup command now needs to recognize when the libraries already exist.
    * To compensate, `mage clean` now removes the `.setup` and `node_modules`  folders as well. A common troubleshooting tactic will be to `mage clean setup` to reinstall all local dependencies

### Build
- `bulid:api` now handles graphql -> typescript generation as well as the go-swagger generation
- bugfix: remove `build:lambda` optimization based on timestamps. Previously, `build:lambda` would stop immediately if the `out/bin` folder had a newer timestamp than any of the source folders - the idea being that if no code changed, we can save several seconds by not running `go build` at all. But if the previous `build:lambda` worked halfway and then ran into an error, the next run would think it was already up to date. To be safe, let's build every time
- `build:opstools` and `build:devtools` has been collapsed to just `build:tools`

### Fmt
After adding the license headers, `gofmt`, `yapf,` `prettier`, and `mage docs` can all run in parallel

Since formatting adds license headers, and `test:ci` fails if any files are changed by formatting, license headers are now enforced in CI (#457)

### Test:CI
Formatting is quite complex - autogenerated docs and swagger files, gofmt, goimports, yapf, license headers, etc. How should we test for it? Instead of trying to diff each tool individually (which is what we currently do), I just run `fmt` itself and then check for diffs in the file hashes. This has two advantages:

- Anytime a developer runs `mage test:ci`, it will *also* format! Formatting mismatches log a warning but will not fail the tests locally (it will fail in CI). This is also faster than running `mage fmt test:ci` like we did before because that was invoking the formatting tools (`gofmt`, `yapf`, etc) twice - once to fmt, once to check formatting
- Changes to `mage fmt` do not require any changes to the test suite

Hashing every source file before and after formatting takes only a few hundred ms. After formatting, all remaining tests can run in parallel. `golangci-lint` now skips `gofmt` and `goimports` (since we're checking those ourselves)

Per developer feedback, I've removed `test:go`, `test:python`, `test:web`, etc. `test:ci` should now be fast enough (~1 min) to check everything. More advanced control could be added in the future through environment variables rather than new mage targets.

Go unit tests now include checks for data races as well (see #656). With the exception of one package (which has been excluded), this doesn't change the test runtime by any noticeable amount.

Finally (@3nvi), `mage test:ci` no longer invokes `npm run validate`, because that includes `prettier` formatting checks which would be redundant. Instead, I add new npm targets to test `eslint` and `tsc` separately

## Testing

- CI (don't worry about the required checks, I will change those to the new values right before merging this PR)
- `mage deploy`

### CODE REVIEW

**Comment by 3nvi** [comment]
@austinbyers Great work on that! Thanks for making things more clear & organized. If possible, please change the title of your PR as it's a bit misleading.

People may think it's related to CircleCI changes, but this fundamentally changes multiple `mage` commands :)

File: docs/gitbook/quick-start.md [hunk 1]
```diff
@@ -62,9 +62,10 @@ Start the development environment:
 Run the following command to deploy Panther:
 
 ```bash
-mage setup:swagger deploy
```
[the remark this answers is not in the document]
Comment (austinbyers) [reply docs/gitbook/quick-start.md:65]: I debated this too, but I prefer the simpler approach:

- If you're deploying from source, you should have all the libraries you need to work with source. It's not clear right now that you need a different setup command to test than you need to deploy. I'm working on having an alternate deployment method this quarter so that you can avoid deploying from source completely
- It doesn't take any longer to do the full setup if you have a reasonable internet connection: it always takes a couple minutes to download all the Go modules you need for compilation on a fresh deploy, so doing the full install in parallel is not appreciably slower

So if it doesn't take longer, it's simpler for everyone to understand, and it enables everyone to run tests immediately before or after deploying, what's the downside?

File: .circleci/config.yml [hunk 2]
```diff
@@ -6,56 +6,31 @@ executors:
...
-    executor: panther-buildpack
-    steps:
-      - setup_frontend
+          name: Install dependencies
+          command: mage setup # TODO: cache these
       - run:
-          name: Lint and audit web source
-          command: mage test:web
+          name: Run all tests
+          command: mage test:ci
```
[the remark this answers is not in the document]
Comment (austinbyers) [reply .circleci/config.yml:21]: I looked into that, but that involves running multiple containers. They can, for example, automatically split `go test` across multiple containers based on file paths, for example.

This is harder to utilize with our monorepo setup - some things can be parallel, some things can't. It's much easier to have full control over the parallelism from a single larger container

The bulk of the runtime in CI is just downloading libraries, which I will fix with caching soon!

File: package.json [hunk 3]
```diff
@@ -124,6 +124,8 @@
     "serve": "node web/scripts/serve.js",
     "graphql-codegen": "graphql-codegen --config web/codegen.yml",
     "graphql-codegen:watch": "npm run graphql-codegen -- --watch",
-    "validate": "eslint --config web/.eslintrc.js \"web/**/*.{ts,js,tsx}\" && prettier --list-different \"web/**/*.{ts,js,tsx,md,json,yml}\" && tsc -p web"
+    "eslint": "eslint --config web/.eslintrc.js \"web/**/*.{ts,js,tsx}\"",
```
[the remark this answers is not in the document]
Comment (austinbyers) [reply package.json:127]: I had to split it so that `mage` can invoke `npm run tsc` and `npm run eslint` separately

[the remark this answers is not in the document]
Comment (austinbyers) [reply package.json:127]: I could invoke node directly, but I'd rather factor it this way so that `npm run validate` and `mage test:ci` are running essentially the same checks.

File: .circleci/config.yml [hunk 4]
```diff
@@ -6,56 +6,31 @@ executors:
...
-  setup_frontend:
-    steps:
-      - checkout
-      - run:
-          name: Install JS dependencies
-          command: mage setup:web
-
 # The pool of jobs that that our CI will be able to run
 jobs:
-  audit_modules:
```
[the remark this answers is not in the document]
Comment (austinbyers) [reply .circleci/config.yml:30]: Yes, that's what `mage test:ci` does now, it happens automatically. That's the main goal of this PR (#457)

**Comment by austinbyers** [comment]
> If possible, please change the title of your PR as it's a bit misleading.

Done!

File: docs/gitbook/quick-start.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (3nvi) [reply docs/gitbook/quick-start.md:65]: I have concerns that have to do with local development mainly, but perhaps they are outside the scope of this. They mainly have to do with installing binaries in different kernels. Running `mage setup` inside docker causes any node modules that have C++ lib dependencies to be installed for the Linux kernel of our docker image (Debian), so when you try and do `npm run start` on your local machine, the command will fail. It will only work inside docker.

In short:

* If you `mage setup:web` in docker, you can only run `npm run start` inside docker and not outside of it
* If  you run `mage setup:web` outside of docker, you can run `npm run start` outside of docker, but not inside of it. 

Now we are "forcing" users to only do local development through Docker. **Deployments are never affected in any way**

[the remark this answers is not in the document]
Comment (austinbyers) [reply docs/gitbook/quick-start.md:65]: I see, it's about development, not deployment. Unless I'm misunderstanding, your concern about npm modules happens regardless of how `mage setup` works - it's about when you run `npm install` at all.

Perhaps the long-term fix is that, at least in my mind, Docker and the local machine shouldn't generally share files. It leads to weird dependency issues (I saw a similar issue with Go modules that we talked about already). People may not even realize that the filesystem is essentially shared. I would prefer if the repo Docker were completely self-contained, requiring no connection to the host filesystem, and then having a command to copy files back and forth when you're done developing

But I've not used Docker much so far for development nor deployment, so I'm not as familiar with your experience with it
